### PR TITLE
chore(js): inbox sdk interface and events improvements

### PR DESCRIPTION
**Pull request by LetItRock**
### What changed? Why was the change needed?
The changes in JS SDK:
- simplified events to two
- changed the names for some user-facing functions like `novu.notifications.list`
- handle the errors inside the public functions and provide the result like `const { data, error }`
- removed api service singleton in favor of inbox service

### Screenshots

<details>
<summary><strong>Expand for optional sections</strong></summary>

### Special notes for your reviewer

</details>

### CODE REVIEW

File: packages/js/src/base-module.ts [hunk 1]
```diff
@@ -3,7 +3,6 @@ import { InboxService } from './api';
 
 import { NovuEventEmitter } from './event-emitter';
 import { Session } from './types';
-import { ApiServiceSingleton } from './utils/api-service-singleton';
```
Comment (LetItRock) [inline packages/js/src/base-module.ts:6]: removed this singleton in favor of `InboxServiceSingleton`

File: packages/js/src/base-module.ts [hunk 2]
```diff
@@ -22,22 +21,22 @@ export class BaseModule {
...
-    this._apiService = ApiServiceSingleton.getInstance();
     this._inboxService = InboxServiceSingleton.getInstance();
-    this._emitter.on('session.initialize.success', ({ result }) => {
-      this.onSessionSuccess(result);
-      this.#callsQueue.forEach(async ({ fn, resolve }) => {
-        resolve(await fn());
-      });
-      this.#callsQueue = [];
-    });
-    this._emitter.on('session.initialize.error', ({ error }) => {
```
Comment (LetItRock) [inline packages/js/src/base-module.ts:34]: the `.success` and `.error` events were consolidated into one `.resolved` that is called with `{ data, error }` props

File: packages/js/src/event-emitter/types.ts [hunk 3]
```diff
@@ -1,104 +1,55 @@
...
-type NovuErrorEvent<A, F = undefined> = Pick<NovuPendingEvent<A>, 'args'> & {
-  error: unknown;
-  fallback?: F;
+type NovuResolvedEvent<A, D> = NovuPendingEvent<A, D> & {
+  error?: unknown;
 };
-// three possible status of the event: pending, success, error
-type EventName<T extends string> = `${T}.pending` | `${T}.success` | `${T}.error`;
+// three possible status of the event: pending, resolved
+type EventName<T extends string> = `${T}.pending` | `${T}.resolved`;
```
Comment (LetItRock) [inline packages/js/src/event-emitter/types.ts:27]: adjusted the types for the events in this file

File: packages/js/src/notifications/helpers.ts [hunk 4]
```diff
@@ -12,7 +13,7 @@ export const read = async ({
   emitter: NovuEventEmitter;
   apiService: InboxService;
   args: ReadArgs;
-}): Promise<Notification> => {
+}): Result<Notification> => {
```
Comment (LetItRock) [inline packages/js/src/notifications/helpers.ts:16]: The `Result` is the type that is promise-based and returns `{ data, error }` props.

File: packages/js/src/notifications/helpers.ts [hunk 5]
```diff
@@ -23,18 +24,19 @@ export const read = async ({
...
+    emitter.emit('notification.read.resolved', { args, data: updatedNotification });
 
-    return updatedNotification;
+    return { data: updatedNotification };
   } catch (error) {
-    emitter.emit('notification.read.error', { args, error });
-    throw error;
+    emitter.emit('notification.read.resolved', { args, error });
+
+    return { error: new NovuError('Failed to read notification', error) };
```
Comment (LetItRock) [inline packages/js/src/notifications/helpers.ts:39]: Custom error objects to provide better error messages. The messages are for the discussion.

File: packages/js/src/notifications/index.ts [hunk 6]
```diff
@@ -1,3 +1,3 @@
-export * from './feeds';
+export * from './notifications';
```
Comment (LetItRock) [inline packages/js/src/notifications/index.ts:1]: Renamed `feeds` to `notifications`, also in the SDK interface.

File: packages/js/src/notifications/notifications.ts [hunk 7]
```diff
@@ -0,0 +1,236 @@
+import { BaseModule } from '../base-module';
```
Comment (LetItRock) [inline packages/js/src/notifications/notifications.ts:1]: This file is what was before the `feeds`.

File: packages/js/src/notifications/notifications.ts [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+  CountArgs,
+  FilterCountResponse,
+  FiltersCountArgs,
+  CountResponse,
+  BaseArgs,
+} from './types';
+import { NovuError } from '../utils/errors';
+
+export class Notifications extends BaseModule {
+  async list({ limit = 10, ...restOptions }: ListNotificationsArgs = {}): Result<ListNotificationsResponse> {
```
Comment (LetItRock) [inline packages/js/src/notifications/notifications.ts:26]: The function to get notifications is now called `list`, also I made similar changes to preferences.

File: packages/js/src/notifications/notifications.ts [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+        this._emitter.emit('notifications.list.resolved', { args, error });
+
+        return { error: new NovuError('Failed to fetch notifications', error) };
+      }
+    });
+  }
+
+  async count(args?: FilterCountArgs): Result<FilterCountResponse>;
+  async count(args?: FiltersCountArgs): Result<FiltersCountResponse>;
+  async count(args: CountArgs): Result<CountResponse> {
```
Comment (LetItRock) [inline packages/js/src/notifications/notifications.ts:56]: The function name was changed to `count`.

File: packages/js/src/notifications/types.ts [hunk 10]
```diff
@@ -1,7 +1,7 @@
 import type { ActionTypeEnum, NotificationFilter } from '../types';
 import { Notification } from './notification';
 
-export type FetchFeedArgs = {
+export type ListNotificationsArgs = {
```
Comment (LetItRock) [inline packages/js/src/notifications/types.ts:4]: Renamed interfaces

File: packages/js/src/ui/api/hooks/useNotifications.ts [hunk 11]
```diff
@@ -0,0 +1,52 @@
+import { createEffect, createSignal } from 'solid-js';
```
Comment (LetItRock) [inline packages/js/src/ui/api/hooks/useNotifications.ts:1]: the file was renamed

File: packages/js/src/base-module.ts [hunk 12]
```diff
@@ -22,36 +22,36 @@ export class BaseModule {
...
+        });
+        this.#callsQueue = [];
+      } else if (error) {
+        this.onSessionError(error);
+        this.#sessionError = error;
+        this.#callsQueue.forEach(({ resolve }) => {
+          resolve({ error: new NovuError('Failed to initialize session', error) });
+        });
+        this.#callsQueue = [];
+      }
```
Comment (SokratisVidros) [inline packages/js/src/base-module.ts:40]:
```else {
  throw new Error("Unknown state, contact support")
}```

Or something like that.

File: packages/js/src/notifications/helpers.ts (hunk 4 above)
Comment (SokratisVidros) [reply packages/js/src/notifications/helpers.ts:16]: 💯

**Review by SokratisVidros — APPROVED** [review]

**Comment by SokratisVidros** [comment]
Nicely done!

File: packages/js/src/event-emitter/types.ts [hunk 13]
```diff
@@ -1,104 +1,55 @@
...
 };
-type NovuErrorEvent<A, F = undefined> = Pick<NovuPendingEvent<A>, 'args'> & {
-  error: unknown;
-  fallback?: F;
+type NovuResolvedEvent<A, D> = NovuPendingEvent<A, D> & {
+  error?: unknown;
 };
-// three possible status of the event: pending, success, error
-type EventName<T extends string> = `${T}.pending` | `${T}.success` | `${T}.error`;
+// three possible status of the event: pending, resolved
```
Comment (BiswaViraj) [inline packages/js/src/event-emitter/types.ts:26]: 🤓 **nitpick (non-blocking):** ‏
```suggestion
// two possible status of the event: pending, resolved
```

File: packages/js/src/ui/api/hooks/useFetchUnreadCount.ts
Comment (BiswaViraj) [inline packages/js/src/ui/api/hooks/useFetchUnreadCount.ts:1]: 💭 **thought (non-blocking):** Should we rename this file to `useUnreadCount`‏

File: packages/js/src/ui/components/elements/Header/MoreActionsOptions.tsx [hunk 14]
```diff
@@ -12,7 +12,7 @@ export const MoreActionsOptions = () => {
       <ActionsItem
         label="Mark all as read"
         /**
-         * TODO: Implement setFeedOptions and isSelected after Filter is implemented
+         * TODO: Implement setNotificationsOptions and isSelected after Filter is implemented
```
Comment (BiswaViraj) [inline packages/js/src/ui/components/elements/Header/MoreActionsOptions.tsx]: ✏ **suggestion:** We can remove this comment, the onClick handler has been implemented now‏

**Review by BiswaViraj — APPROVED** [review]

File: packages/js/src/event-emitter/types.ts (hunk 13 above)
Comment (LetItRock) [reply packages/js/src/event-emitter/types.ts:26]: I've already updated that but in a different PR as I couldn't commit the changes to this one as I had many other changes in there.

File: packages/js/src/base-module.ts (hunk 12 above)
Comment (LetItRock) [reply packages/js/src/base-module.ts:40]: This is the specific case when the session fails to initialize, I'll update it to: `Failed to initialize session, please contact the support`

File: packages/js/src/ui/api/hooks/useFetchUnreadCount.ts
Comment (LetItRock) [reply packages/js/src/ui/api/hooks/useFetchUnreadCount.ts:1]: It's not used, I'll remove it.

**Pull request merged**